tools/src/wyvern/target/corewyvernIL/decl/DefDeclaration.java

```diff
@@ -79,4 +80,20 @@ public DeclType typeCheck(TypeContext ctx, TypeContext thisCtx) {
 			throw new RuntimeException("body doesn't match declared type");
 		return new DefDeclType(getName(), type, formalArgs);
 	}
+
+	@Override
+	public Set<String> getFreeVariables() {
+		
+		// Get all free variables in the body of the method.
+		Set<String> freeVars = body.getFreeVariables();
+		
+		// Remove variables that became bound in this method's scope.
+		freeVars.remove(this.getName());
```

Comment:
Actually the name of the definition is not bound in the body --- at the level of the Core IL, recursive calls must be of the form "this.mymethod(...)" not just "mymethod(...)" [there is a sugar that makes this work at the source level but it goes away in translation to the core]